documentation/src/main/asciidoc/mapper-orm-bridge-index-field-dsl.asciidoc

```diff
@@ -165,3 +165,105 @@ and pass it to the bridge for later use.
+The field, then, is considered _dynamic_.
+
+Dynamic fields are not declared on bootstrap,
+but need to match a field _template_ that is declared on bootstrap.
+The template includes the field type and structural information (multi-valued or not, ...),
+but omits the field name.
+
+Field templates are declared with the `fieldTemplate` methods, as shown below.
+
+.Declaring a field template
+====
+[source, JAVA, indent=0, subs="+callouts"]
+----
+include::{sourcedir}/org/hibernate/search/documentation/mapper/orm/bridge/document/model/dsl/dynamic/UserMetadataBinder.java[tags=bind]
+----
+<1> Declare an object field with `objectField`.
+It's better to always host your dynamic fields on a dedicated object field,
+to avoid conflicts with other templates.
+<2> Declare a field template with `fieldTemplate`.
+<3> Pass the *template* name -- this is not the field name, and is only used to uniquely identify the template.
+<4> Define the field type.
+<5> On contrary to static field declarations, field template declarations do not return a field reference,
+because you won't need it when writing to the document.
+<6> Get a reference to the declared object field
+and pass it to the bridge for later use.
+====
+
+[NOTE]
+====
+Though rarely necessary, you can also declare templates for object fields using the `objectFieldTemplate` methods.
+====
+
+Assuming a field template was declared,
+the bridge can then add dynamic fields to the document when indexing,
+as shown in the example below.
+
+.Writing to a dynamic field
+====
+[source, JAVA, indent=0, subs="+callouts"]
+----
+include::{sourcedir}/org/hibernate/search/documentation/mapper/orm/bridge/document/model/dsl/dynamic/UserMetadataBinder.java[tags=write]
+----
+<1> Add an object to the `userMetadata` object field for the current document,
+and get a reference to that object.
+<2> Add one field per user metadata entry,
+with the field name and field value defined by the user.
+Note that field names should usually be validated before that point,
+in order to avoid exotic characters (whitespaces, dots, ...).
+====
+
+It is also possible to add multiple fields with different types to the same object.
+To that end, make sure that the format of a field can be inferred from the field name.
+You can then declare multiple templates and assign a path pattern to each template,
+as shown below.
+
+.Declaring multiple field templates with different types
+====
+[source, JAVA, indent=0, subs="+callouts"]
+----
+include::{sourcedir}/org/hibernate/search/documentation/mapper/orm/bridge/document/model/dsl/dynamic/MultiTypeUserMetadataBinder.java[tags=bind]
+----
+<1> Declare an object field with `objectField`.
+<2> Declare a field template for integer with `fieldTemplate`.
+<3> Pass the *template* name.
+<4> Define the field type as integer, sortable.
+<5> Assign a path pattern to the template, so that only fields ending with `_int` will be considered as integers.
```

Comment:
I like the idea of having templates. I don't know but maybe we should find a way to make the association to the field more dynamic, providing the template within the field name, e.g. having in the map key something like `fieldName:templateName`.

Moreover, I would let the user to define templates globally, for the entire application, not only for a given property.

Reply:
> I like the idea of having templates. I don't know but maybe we should find a way to make the association to the field more dynamic, providing the template within the field name, e.g. having in the map key something like `fieldName:templateName`.

Passing a reference to the template would indeed be nice, and that's I wanted to do initially, but that's simply not possible with Elasticsearch. You can't specify the type at indexing time, you have to rely on a template with a name pattern.

Eventually we might consider introducing an extension for Lucene only, I suppose. But I'd rather wait until we get a real use case.

> Moreover, I would let the user to define templates globally, for the entire application, not only for a given property.

I don't think it would be a good idea to do that for the entire application. Templates are defined per-index in Elasticsearch, and I'd rather stick to that so that changing templates will only affect the indexes that actually need this template.

We should keep in mind that dynamic fields are not generally necessary. They are not necessary when you directly map properties of your entities, because then you can simply use `@*Field` annotations and create "static" fields. They are mostly necessary for advanced use cases such as mapping a `Map` or a JSON object whose schema is not known in advance.

If you need to define templates for the whole index, that's totally possible, as explained below:

> Templates declared on a given schema element can be matched in children of that element.
> For example, if you declare templates at the root of your entity (through a <<mapper-orm-bridge-typebridge,type bridge>>),
> these templates will be implicitly available in every single property bridge of that entity.
> In such cases, templates declared in property bridges will take precedence over those declared in the type bridge.